services/brig/src/Brig/API/User.hs

```diff
@@ -875,44 +875,68 @@ lookupProfile :: UserId -> UserId -> AppIO (Maybe UserProfile)
 lookupProfile self other = listToMaybe <$> lookupProfiles self [other]
 
 -- | Obtain user profiles for a list of users as they can be seen by
--- a given user 'A'. User 'A' can see the 'FullProfile' of any other user 'B',
--- if the reverse relation (B -> A) is either 'Accepted' or 'Sent'.
--- Otherwise only the 'PublicProfile' is accessible for user 'A'.
+-- a given user 'self'. User 'self' can see the 'FullProfile' of any other user 'other',
+-- if the reverse relation (other -> self) is either 'Accepted' or 'Sent'.
+-- Otherwise only the 'PublicProfile' is accessible for user 'self'.
+-- If 'self' is an unknown 'UserId', return '[]'.
 lookupProfiles ::
-  -- | User 'A' on whose behalf the profiles are requested.
+  -- | User 'self' on whose behalf the profiles are requested.
   UserId ->
-  -- | The users ('B's) for which to obtain the profiles.
+  -- | The users ('others') for which to obtain the profiles.
   [UserId] ->
   AppIO [UserProfile]
 lookupProfiles self others = do
   users <- Data.lookupUsers others >>= mapM userGC
   css <- toMap <$> Data.lookupConnectionStatus (map userId users) [self]
   emailVisibility' <- view (settings . emailVisibility)
-  return $ map (toProfile emailVisibility' css) users
+  mapM (toProfile emailVisibility' css) users
   where
     toMap :: [ConnectionStatus] -> Map UserId Relation
     toMap = Map.fromList . map (csFrom &&& csStatus)
-    toProfile :: EmailVisibility -> Map UserId Relation -> User -> UserProfile
-    toProfile emailVisibility' css u =
+    --
+    toProfile :: EmailVisibility -> Map UserId Relation -> User -> AppIO UserProfile
+    toProfile emailVisibility' css u = do
       let cs = Map.lookup (userId u) css
-          profileEmail' = getEmailForProfile u emailVisibility'
           baseProfile =
             if userId u == self || cs == Just Accepted || cs == Just Sent
               then connectedProfile u
               else publicProfile u
-       in baseProfile {profileEmail = profileEmail'}
+      profileEmail' <- getEmailForProfile u self emailVisibility'
+      pure baseProfile {profileEmail = profileEmail'}
 
 -- | Gets the email if it's visible to the requester according to configured settings
 getEmailForProfile ::
-  -- | The user who's profile is being requested
   User ->
+  UserId ->
   EmailVisibility ->
-  Maybe Email
-getEmailForProfile _ EmailVisibleToSelf = Nothing
-getEmailForProfile u EmailVisibleIfOnTeam =
-  if isJust (userTeam u)
-    then userEmail u
-    else Nothing
+  AppIO (Maybe Email)
+getEmailForProfile _ _ EmailVisibleToSelf = pure Nothing
+getEmailForProfile profileOwner _ EmailVisibleIfOnTeam =
+  pure $
+    if isJust (userTeam profileOwner)
+      then userEmail profileOwner
+      else Nothing
+getEmailForProfile profileOwner profileViewer EmailVisibleIfOnSameTeam = do
+  viewerUser <- getViewerUser
```

Comment:
Do I understand this correctly? We fetch the user from `brig`, and then check `galley` for the team membership?

We can go straight to `galley` and be done with it - I think there's no endpoint for that in `galley` right now but we should add an internal endpoint in the lines of `getBindingTeamMembership` which would simplify things here (and avoid extra DB lookups).

Also: `getViewerUser` is getting called for every "other profile" we look at, right? So we're doing the same thing multiple times or am I missing something?

Reply:
> we're doing the same thing multiple times or am I missing something?

Yes, that's true, I sketched out an idea to do this as we want (only do the call if we need to, never more than once), but there might be nicer ways.